feature-changeaddress/src/main/kotlin/com/example/feature/changeaddress/data/ChangeAddressRepository.kt

```diff
@@ -0,0 +1,153 @@
+package com.example.feature.changeaddress.data
+
+import arrow.core.Either
+import arrow.core.continuations.either
+import com.apollographql.apollo3.ApolloClient
+import com.apollographql.apollo3.api.Optional
+import com.hedvig.android.apollo.safeExecute
+import com.hedvig.android.apollo.toEither
+import com.hedvig.android.core.common.ErrorMessage
+import octopus.MoveIntentCommitMutation
+import octopus.MoveIntentCreateMutation
+import octopus.MoveIntentRequestMutation
+import octopus.type.MoveApartmentSubType
+import octopus.type.MoveIntentCreateInput
+import octopus.type.MoveToAddressInput
+import octopus.type.MoveToApartmentInput
+
+internal interface ChangeAddressRepository {
+  suspend fun createMoveIntent(): Either<ErrorMessage, MoveIntent>
+  suspend fun createQuotes(id: MoveIntentId, input: MoveIntentRequestInput): Either<ErrorMessage, List<MoveQuote>>
+  suspend fun commitMove(id: MoveIntentId): Either<ErrorMessage, MoveResult>
+}
+
+internal class NetworkChangeAddressRepository(
+  private val apolloClient: ApolloClient,
+) : ChangeAddressRepository {
+  override suspend fun createMoveIntent(): Either<ErrorMessage, MoveIntent> {
+    return either {
+      val result = apolloClient
+        .mutation(MoveIntentCreateMutation(MoveIntentCreateInput(toType = octopus.type.MoveToType.APARTMENT)))
+        .safeExecute()
+        .toEither(::ErrorMessage)
+        .bind()
+        .moveIntentCreate
+
+      val moveIntent = result.moveIntent
+      val userError = result.userError
+
+      when {
+        userError != null -> shift(ErrorMessage(userError.message))
+        moveIntent != null -> moveIntent.toMoveIntent()
+        else -> shift(ErrorMessage("No data found in MoveIntent"))
+      }
```

Comment:
Yes perfect!

Reply:
A question, since the backend returns these two fields separately, there isn't any guarantee in the type at least that only one of the two is going to be non-null, or that neither of them are going to be non-null. Despite that, have they communicated such a guarantee?

For example in this implementation, if the response contains both an error and a moveIntent response, we are going to be using only the error. Or should that be an impossible return value from them anyway?

Reply:
We just have to assume that both will be non-null. The purchase journey team talked about modelling this with return types so that the schema got more expressive. It got too complex in the end though and they ended up with this design instead. Ill check if the moveIntent is null first though, that makes more sense.

Reply:
Alright, let's go with that then yeah